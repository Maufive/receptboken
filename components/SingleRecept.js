import React, { Component } from "react";
import Router from "next/router";
import Head from "next/head";
import axios from "axios";
import Moment from "react-moment";
import { MessageConsumer } from "./providers/MessageProvider";
import Stars from "./Stars";
import { Loading } from "./Loading";
import Author from "./Author";
import Heart from "./Heart";
import ClockIcon from "../svg/clock.svg";
import TrashIcon from "../svg/trash.svg";
import TagIcon from "../svg/price-tag.svg";
import CalenderIcon from "../svg/calendar.svg";
import DishIcon from "../svg/dishes.svg";
import {
	Wrapper,
	DetailsBar,
	ImageAndTags,
	Tag,
	Ingredienser,
	Beskrivning,
	Container,
	IconContainer,
	DeleteContainer,
	ImgContainer,
	Dot
} from "./styles/ReceptStyles";
import AddToShoppinglist from "./AddToShoppinglist";

class SingleRecept extends Component {
	state = {
		recept: null,
		loading: false,
		error: null,
		owner: false
	};

	async componentDidMount() {
		await this.getData();
		if (this.props.user) {
			this.verifyUser();
		}
	}

	getData = async () => {
		await this.setState({ loading: true });
		await axios
			.get(`https://receptboken.herokuapp.com/recipe/one/${this.props.id}`)
			.then(response => {
				this.setState({ recept: response.data, loading: false });
			})
			.catch(error => {
				console.log(error);
				Router.push("/");
			});
	};

	// Om användaren blivit verifierad och idt stämmer överens med idt som skapat receptet, sätt owner till boolean från api't.
	// och visa en knapp för att ta bort receptet.
	verifyUser = () => {
		axios
			.post(
				`https://receptboken.herokuapp.com/auth/verify/author/${
					this.state.recept.author
				}`
			)
			.then(response => {
				this.setState({ owner: response.data });
			})
			.catch(error => console.log(error));
	};

	deleteRecipe = async e => {
		e.preventDefault();
		await this.setState({ loading: true });
		await axios
			.post(
				`https://receptboken.herokuapp.com/recipe/delete/${this.props.id}`,
				{
					user: this.props.user
				}
			)
			.then(response => {
				this.setState({ loading: false });
				Router.push("/");
			})
			.catch(error => {
				console.log(error);
			});
	};

	render() {
		const { recept } = this.state;
		if (!this.state.recept) return <Loading />;
		return (
			<Wrapper>
				<h1>{recept.title}</h1>
				<Head>
					<title>Receptboken | {recept.title}</title>
				</Head>

				<DetailsBar>
					<Stars id={this.props.id} />
					<MessageConsumer>
						{({ setMessage }) => (
							<Heart
								id={recept._id}
								setMessage={setMessage}
								user={this.props.user}
								setUser={this.props.setUser}
							/>
						)}
					</MessageConsumer>
					<IconContainer>
						<ClockIcon />
						{recept.timeRequired.toString()}
					</IconContainer>
					<IconContainer>
						<DishIcon />
						{recept.servings}
					</IconContainer>
				</DetailsBar>

				<ImageAndTags>
					<ImgContainer>
						<img src={recept.photo} alt="Foto på maten" />
					</ImgContainer>
					<div>
						<div>
							{recept.tags.map(tag => (
								<Tag key={tag}>
									<TagIcon />
									<p>{tag}</p>
								</Tag>
							))}
						</div>

						{this.props.user && (
							<div>
								<MessageConsumer>
									{({ setMessage }) => (
										<AddToShoppinglist
											userid={this.props.user._id}
											receptid={recept._id}
											ingredients={recept.ingredients}
											setMessage={setMessage}
										/>
									)}
								</MessageConsumer>
							</div>
						)}

						<Author id={recept.author} />
						<IconContainer>
							<CalenderIcon />
							<Moment format="YYYY/MM/DD">{recept.created}</Moment>
						</IconContainer>

						{this.state.owner && (
							<DeleteContainer onClick={this.deleteRecipe}>
								<TrashIcon />
								Ta bort recept
							</DeleteContainer>
						)}
					</div>
				</ImageAndTags>

				<Container>
					<h2>Du Behöver:</h2>
					<Ingredienser>
						{recept.ingredients.map(ingredient => (
							<div key={ingredient.input}>
								<Dot />
								<p>
									{ingredient.numberOfUnits} {ingredient.units}{" "}
									{ingredient.input}
								</p>
							</div>
						))}
					</Ingredienser>

					<h2>Gör så här:</h2>
					<Beskrivning>
						{recept.description.map(step => (
							<div key={step}>
								<Dot />
								<p>{step}</p>
							</div>
						))}
					</Beskrivning>
				</Container>
			</Wrapper>
		);
	}
}

export default SingleRecept;
