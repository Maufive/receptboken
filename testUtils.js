export const fakeRecept = {
	_id: "1234",
	author: "5678436374re67fwuf76",
	created: "Idag",
	description: ["Steg1", "Steg2", "Steg3"],
	id: "1337",
	ingredients: [
		{
			input: "Lök",
			numberOfUnits: "2",
			units: "dl"
		}
	],
	photo: "asdf123.jpg",
	reviews: [{ score: 1 }],
	servings: "2",
	tags: ["Mat", "Gott"],
	timeRequired: "20 min",
	title: "Mitt jättegoda recept"
};

export const fakeUser = {
	_id: "1234",
	username: "fakeuser@gmail.com",
	password: "1234",
	hearts: [],
	fname: "Niklas",
	lname: "Albinsson",
	description: "Hej jag heter Niklas",
	photo: "http://fakephoto.jpg"
};

export const fakeList = {
	list: [
		{
			input: "Lök",
			numberOfUnits: "5",
			units: "dl"
		},
		{
			input: "Tomater",
			numberOfUnits: "10",
			units: "st"
		}
	]
};
