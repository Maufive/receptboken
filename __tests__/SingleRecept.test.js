import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import SingleRecept from "../components/SingleRecept";
import { fakeRecept } from "../testUtils";

describe("<SingleRecept />", () => {
	let wrapper;
	beforeEach(() => (wrapper = shallow(<SingleRecept />)));

	it("shows Loader when there is no state", () => {
		expect(wrapper.exists("Loading")).toBe(true);
	});

	it("Renders out after setting state", () => {
		wrapper.setState({ recept: fakeRecept });
		expect(toJSON(wrapper)).toMatchSnapshot();
		expect(wrapper.find("h1").text()).toBe(fakeRecept.title);
		expect(wrapper.find("img").prop("src")).toEqual(fakeRecept.photo);
	});
});
