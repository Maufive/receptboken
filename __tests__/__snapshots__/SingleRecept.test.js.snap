// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleRecept /> Renders out after setting state 1`] = `
<ReceptStyles__Wrapper>
  <h1>
    Mitt jättegoda recept
  </h1>
  <SideEffect(Head)>
    <title>
      Receptboken | 
      Mitt jättegoda recept
    </title>
  </SideEffect(Head)>
  <ReceptStyles__DetailsBar>
    <Stars />
    <ContextConsumer>
      <Component />
    </ContextConsumer>
    <ReceptStyles__IconContainer>
      <ClockIcon
        viewBox="0 0 60 60"
        xmlns="http://www.w3.org/2000/svg"
      />
      20 min
    </ReceptStyles__IconContainer>
    <ReceptStyles__IconContainer>
      <DishIcon
        viewBox="0 0 49.749 49.749"
        xmlns="http://www.w3.org/2000/svg"
      />
      2
    </ReceptStyles__IconContainer>
  </ReceptStyles__DetailsBar>
  <ReceptStyles__ImageAndTags>
    <ReceptStyles__ImgContainer>
      <img
        alt="Foto på maten"
        src="asdf123.jpg"
      />
    </ReceptStyles__ImgContainer>
    <div>
      <div>
        <ReceptStyles__Tag
          key="Mat"
        >
          <TagIcon
            viewBox="0 0 56.414 56.414"
            xmlns="http://www.w3.org/2000/svg"
          />
          <p>
            Mat
          </p>
        </ReceptStyles__Tag>
        <ReceptStyles__Tag
          key="Gott"
        >
          <TagIcon
            viewBox="0 0 56.414 56.414"
            xmlns="http://www.w3.org/2000/svg"
          />
          <p>
            Gott
          </p>
        </ReceptStyles__Tag>
      </div>
      <Author
        id="5678436374re67fwuf76"
      />
      <ReceptStyles__IconContainer>
        <CalenderIcon
          viewBox="0 0 60 60"
          xmlns="http://www.w3.org/2000/svg"
        />
        <t
          ago={false}
          calendar={false}
          decimal={false}
          element="time"
          filter={[Function]}
          format="YYYY/MM/DD"
          fromNow={false}
          interval={60000}
          onChange={[Function]}
          titleFormat=""
          toNow={false}
          unit={null}
          unix={false}
          utc={false}
          withTitle={false}
        >
          Idag
        </t>
      </ReceptStyles__IconContainer>
    </div>
  </ReceptStyles__ImageAndTags>
  <ReceptStyles__Container>
    <h2>
      Du Behöver:
    </h2>
    <ReceptStyles__Ingredienser>
      <div
        key="Lök"
      >
        <ReceptStyles__Dot />
        <p>
          2
           
          dl
           
          Lök
        </p>
      </div>
    </ReceptStyles__Ingredienser>
    <h2>
      Gör så här:
    </h2>
    <ReceptStyles__Beskrivning>
      <div
        key="Steg1"
      >
        <ReceptStyles__Dot />
        <p>
          Steg1
        </p>
      </div>
      <div
        key="Steg2"
      >
        <ReceptStyles__Dot />
        <p>
          Steg2
        </p>
      </div>
      <div
        key="Steg3"
      >
        <ReceptStyles__Dot />
        <p>
          Steg3
        </p>
      </div>
    </ReceptStyles__Beskrivning>
  </ReceptStyles__Container>
</ReceptStyles__Wrapper>
`;
